### ISSUE TITLE
Test failures on OS X

### ISSUE THREAD

**Issue by msabramo** [Agamnentzar/ag-psd#6]
Not sure why but several tests fail on OS X because the file sizes aren't what it's expecting. When I run it in a Docker container, all the tests pass.

```
  12) PsdWriter
       writes PSD file (layer-mask):
     Error: Buffers differ in size actual: 29591 expected: 29585 (ArrayBufferPsdWriter)
      at Object.compareBuffers (src/test/common.ts:149:9)
      at Context.<anonymous> (src/test/psdWriter.spec.ts:348:4)

  13) PsdWriter
       writes PSD file (layer-offsets):
     Error: Buffers differ in size actual: 20053 expected: 20031 (ArrayBufferPsdWriter)
      at Object.compareBuffers (src/test/common.ts:149:9)
      at Context.<anonymous> (src/test/psdWriter.spec.ts:348:4)

  14) PsdWriter
       writes PSD file (no-background):
     Error: Buffers differ in size actual: 119561 expected: 119537 (ArrayBufferPsdWriter)
      at Object.compareBuffers (src/test/common.ts:149:9)
      at Context.<anonymous> (src/test/psdWriter.spec.ts:348:4)

  15) PsdWriter
       writes PSD file (simple):
     Error: Buffers differ in size actual: 84299 expected: 84281 (ArrayBufferPsdWriter)
      at Object.compareBuffers (src/test/common.ts:149:9)
      at Context.<anonymous> (src/test/psdWriter.spec.ts:348:4)
```

**Comment by Agamnentzar** [comment]
The test file should be saved in `results/simple.psd` can you send me that file so I can check the diff ?

**Comment by msabramo** [comment]
Here you go. I hope this helps:

[simple.psd.zip](https://github.com/Agamnentzar/ag-psd/files/2975416/simple.psd.zip)

Thanks for looking!

**Comment by Agamnentzar** [comment]
Thumbnail data relies on jpeg compression and it seems that in this situation the native OS X jpeg compressor produces slightly different results. It's not an issue, but I'll have to figure out how to update the tests to work around that issue.

**Comment by msabramo** [comment]
Ah makes sense. I wonder if it’s possible to set the JPEG compression level to 0 or make it use another format for the tests like PNG, webp, etc.?

**Comment by msabramo** [comment]
Maybe use something like this?

```diff
diff --git a/src/test/common.ts b/src/test/common.ts
index 4b22a20..0a33d33 100644
--- a/src/test/common.ts
+++ b/src/test/common.ts
@@ -99,7 +99,7 @@ export function extractPSD(filePath: string, psd: Psd) {

 export function saveCanvas(fileName: string, canvas: HTMLCanvasElement | undefined) {
        if (canvas) {
-               fs.writeFileSync(fileName, canvas.toBuffer());
+               fs.writeFileSync(fileName, canvas.toBuffer('image/jpeg', { quality: 1.0 }));
        }
 }
```

**Comment by Agamnentzar** [comment]
It's not the test canvas saving, it's thumbnail generation here: https://github.com/Agamnentzar/ag-psd/blob/master/src/imageResources.ts#L350

**Comment by msabramo** [comment]
Ah okay. Same idea might be possible here since `toDataURL` seems to take options for specifying compression.

From https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/toDataURL

<img width="705" alt="Screen Shot 2019-03-17 at 12 21 40 PM" src="https://user-images.githubusercontent.com/305268/54496666-3f4d0680-48af-11e9-87c1-d8b58dff16e7.png">

**Comment by msabramo** [comment]
```diff
diff --git a/src/imageResources.ts b/src/imageResources.ts
index ad6a66f..49bb48a 100644
--- a/src/imageResources.ts
+++ b/src/imageResources.ts
@@ -347,7 +347,7 @@ addHandler(
        },
        (writer, target) => {
                const thumb = target.thumbnail!;
-               const data = toByteArray(thumb.toDataURL('image/jpeg').substr('data:image/jpeg;base64,'.length));
+               const data = toByteArray(thumb.toDataURL('image/jpeg', 1.0).substr('data:image/jpeg;base64,'.length));
                const bitsPerPixel = 24;
                const widthBytes = (thumb.width * bitsPerPixel + 31) / 32 * 4;
                const planes = 1;
```

?

**Comment by Agamnentzar** [comment]
Yeah, that could work, can you check it on the system where it was failing ?

**Comment by msabramo** [comment]
I can send you the resulting file I get on my Mac with that change and then you can compare it with what you get. Or I guess I could compare with results from a Linux Docker container.

**Comment by Agamnentzar** [comment]
oh right, would have to also generate new test files first, can you generate new simple.psd using this change on the system where is was failing and send me that one.

**Comment by Agamnentzar** [comment]
Actually when I change it to 1 it produces the same results for me, so you just need to check if it passes the tests after change

**Comment by msabramo** [comment]
Hmmm, still failing. :-(

[simple.psd.zip](https://github.com/Agamnentzar/ag-psd/files/2975712/simple.psd.zip)

**Comment by msabramo** [comment]
Here are the image sizes that I get with some added `console.log` statements:

```
images[i.name] =  [Canvas 160x107]  -  13603
i.canvas =  [Canvas 160x107]  -  13133
!․․images[i.name] =  [Canvas 160x107]  -  14679
i.canvas =  [Canvas 160x107]  -  14747
!images[i.name] =  [Canvas 28x19]  -  1131
i.canvas =  [Canvas 28x19]  -  1146
!․․images[i.name] =  [Canvas 160x107]  -  35065
i.canvas =  [Canvas 160x107]  -  35368
!․․images[i.name] =  [Canvas 9x14]  -  455
i.canvas =  [Canvas 9x14]  -  451
!․images[i.name] =  [Canvas 160x160]  -  22911
i.canvas =  [Canvas 160x160]  -  22787
!images[i.name] =  [Canvas 160x106]  -  9080
i.canvas =  [Canvas 160x106]  -  8764
!images[i.name] =  [Canvas 160x107]  -  24225
i.canvas =  [Canvas 160x107]  -  24472
!․․images[i.name] =  [Canvas 100x100]  -  7091
i.canvas =  [Canvas 100x100]  -  7072
```

Here are the `console.log`s that I added in case you want to compare:

```diff
diff --git a/src/test/psdReader.spec.ts b/src/test/psdReader.spec.ts
index 9f8772e..e8d0ca5 100644
--- a/src/test/psdReader.spec.ts
+++ b/src/test/psdReader.spec.ts
@@ -89,7 +89,14 @@ describe('PsdReader', () => {
                        clearEmptyCanvasFields(expected);

                        expect(psd).eql(expected, f);
-                       compare.forEach(i => compareCanvases(images[i.name], i.canvas, `${f}/${i.name}`));
+                       compare.forEach(i => {
+                               // console.log('images[i.name] = ', images[i.name]);
+                               if (i.canvas && images[i.name].toBuffer().length != i.canvas.toBuffer().length) {
+                                       console.log('images[i.name] = ', images[i.name], ' - ', images[i.name].toBuffer().length);
+                                       console.log('i.canvas = ', i.canvas, ' ', i.canvas.toBuffer().length);
+                               }
+                               compareCanvases(images[i.name], i.canvas, `${f}/${i.name}`)
+                       });
```

**Comment by msabramo** [comment]
Here are all the results:

[results.zip](https://github.com/Agamnentzar/ag-psd/files/2975730/results.zip)

**Comment by msabramo** [comment]
whereas if I run the tests in a (Debian GNU/Linux 9 (stretch)) `node:8` Docker container, I get:

```
$ docker run -it -v $(pwd):/pwd -w /pwd node:8 bash -c 'npm install && npm test'

> canvas@2.3.1 install /pwd/node_modules/canvas
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Using needle for node-pre-gyp https download
[canvas] Success: "/pwd/node_modules/canvas/build/Release/canvas.node" is installed via remote
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.
npm WARN lifecycle ag-psd@6.1.0~prepublish: cannot run in wd ag-psd@6.1.0 npm run build (wd=/pwd)
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 653 packages from 994 contributors and audited 9268 packages in 72.629s
found 1 moderate severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details

> ag-psd@6.1.0 test /pwd
> gulp test

[21:32:47] Using gulpfile /pwd/gulpfile.js
[21:32:47] Starting 'test'...
[21:32:47] Starting 'build-js'...
[21:32:53] Finished 'build-js' after 5.87 s
[21:32:53] Starting 'tests'...


  ․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․

  74 passing (3s)
  1 pending

[21:33:05] Finished 'tests' after 12 s
[21:33:05] Finished 'test' after 18 s
```

**Comment by msabramo** [comment]
Interestingly, output with #6 is showing me that the actual bytes differ quite a bit, but the rendered images look the same to my eyes. See below - left is the expected image and right is the image that my Mac spit out:

![Screen Shot 2019-03-18 at 4 09 17 PM](https://user-images.githubusercontent.com/305268/54569549-819e4280-4998-11e9-81ae-d2ce935dcce3.png)

**Comment by Agamnentzar** [comment]
I realized that the thumbnail is also dependent on image scaling, so different graphics library will also skew the result, so I decided to disable thumbnails for test data.

Fixed in 6a04af0e5d6c3508f9d0bf2cb0b6e84acd54def9

**Comment by msabramo** [comment]
6a04af0 seems to have fixed the failing `PsdWriter` tests, but the `PsdReader` tests are still failing:

```
$ npm test

> ag-psd@6.2.0 test /Users/abramowi/dev/git-repos/ag-psd
> gulp test

[16:45:07] Using gulpfile ~/dev/git-repos/ag-psd/gulpfile.js
[16:45:07] Starting 'test'...
[16:45:07] Starting 'build-js'...
[16:45:10] Finished 'build-js' after 3.09 s
[16:45:10] Starting 'tests'...


  ․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․,․!․․!!․․!․․!․!!!․․!․․․․․․․․․․․․․․․․․․․․․․

  65 passing (1s)
  2 pending
  9 failing

  1) PsdReader
       reads PSD file (background):
     Error: Actual canvas (68480 bytes) different than expected (background/thumb.png: 68480 bytes) at index 28: actual 250 vs. expected 249
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  2) PsdReader
       reads PSD file (blend-mode):
     Error: Actual canvas (68480 bytes) different than expected (blend-mode/thumb.png: 68480 bytes) at index 129: actual 251 vs. expected 253
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  3) PsdReader
       reads PSD file (cat):
     Error: Actual canvas (2128 bytes) different than expected (cat/thumb.png: 2128 bytes) at index 38: actual 252 vs. expected 250
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  4) PsdReader
       reads PSD file (groups):
     Error: Actual canvas (68480 bytes) different than expected (groups/thumb.png: 68480 bytes) at index 60: actual 255 vs. expected 254
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  5) PsdReader
       reads PSD file (lantern):
     Error: Actual canvas (504 bytes) different than expected (lantern/thumb.png: 504 bytes) at index 0: actual 248 vs. expected 249
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  6) PsdReader
       reads PSD file (layer-mask):
     Error: Actual canvas (102400 bytes) different than expected (layer-mask/thumb.png: 102400 bytes) at index 129: actual 255 vs. expected 254
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  7) PsdReader
       reads PSD file (layer-offsets-read):
     Error: Actual canvas (67840 bytes) different than expected (layer-offsets-read/thumb.png: 67840 bytes) at index 9862: actual 255 vs. expected 253
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  8) PsdReader
       reads PSD file (layers):
     Error: Actual canvas (68480 bytes) different than expected (layers/thumb.png: 68480 bytes) at index 61: actual 137 vs. expected 138
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)

  9) PsdReader
       reads PSD file (section-2):
     Error: Actual canvas (40000 bytes) different than expected (section-2/thumb.png: 40000 bytes) at index 12400: actual 144 vs. expected 146
      at Object.compareCanvases (src/test/common.ts:144:10)
      at /Users/abramowi/dev/git-repos/ag-psd/src/test/psdReader.spec.ts:92:25
      at Array.forEach (<anonymous>)
      at Context.<anonymous> (src/test/psdReader.spec.ts:92:12)



Mocha exited with code 9
[16:45:13] Finished 'tests' after 2.68 s
[16:45:13] Finished 'test' after 5.78 s
```

With #10, everything passes for me though on OS X:

```
$ npm test

> ag-psd@6.2.0 test /Users/abramowi/dev/git-repos/ag-psd
> gulp test

[16:45:40] Using gulpfile ~/dev/git-repos/ag-psd/gulpfile.js
[16:45:40] Starting 'test'...
[16:45:40] Starting 'build-js'...
[16:45:43] Finished 'build-js' after 3.11 s
[16:45:43] Starting 'tests'...


  ․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․

  74 passing (1s)
  2 pending

[16:45:45] Finished 'tests' after 2.48 s
[16:45:45] Finished 'test' after 5.6 s
```

and also in Linux:

```
$ docker run -it -v $(pwd):/pwd -w /pwd node:8 bash -c 'npm install && npm test'

> canvas@2.3.1 install /pwd/node_modules/canvas
> node-pre-gyp install --fallback-to-build

node-pre-gyp WARN Using needle for node-pre-gyp https download
[canvas] Success: "/pwd/node_modules/canvas/build/Release/canvas.node" is installed via remote
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.
npm WARN lifecycle ag-psd@6.2.0~prepublish: cannot run in wd ag-psd@6.2.0 npm run build (wd=/pwd)
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 656 packages from 1008 contributors and audited 9272 packages in 70.929s
found 1 moderate severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details

> ag-psd@6.2.0 test /pwd
> gulp test

[23:47:28] Using gulpfile /pwd/gulpfile.js
[23:47:28] Starting 'test'...
[23:47:28] Starting 'build-js'...
[23:47:34] Finished 'build-js' after 6.33 s
[23:47:34] Starting 'tests'...


  ․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․

  74 passing (3s)
  2 pending

[23:47:43] Finished 'tests' after 9.24 s
[23:47:43] Finished 'test' after 16 s
```

**Comment by Agamnentzar** [comment]
Forgot that it's an issue for reading too, should be fixed in a907b7b8aaf5c3059b5934defac26c532ef2611e

**Comment by msabramo** [comment]
👍 

```
$ git up
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 5 (delta 0), reused 1 (delta 0), pack-reused 0
Unpacking objects: 100% (5/5), done.
From github.com:Agamnentzar/ag-psd
   8737347..a907b7b  master     -> origin/master
Updating 8737347..a907b7b
Fast-forward
 src/test/psdReader.spec.ts | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)
Current branch master is up to date.
a907b7b (HEAD -> master, origin/master, origin/HEAD) Remove checking thumbnail canvases in reading tests


abramowi at marcs-mbp in ~/dev/git-repos/ag-psd (master)
$ npm test

> ag-psd@6.2.0 test /Users/abramowi/dev/git-repos/ag-psd
> gulp test

[17:15:03] Using gulpfile ~/dev/git-repos/ag-psd/gulpfile.js
[17:15:03] Starting 'test'...
[17:15:03] Starting 'build-js'...
[17:15:06] Finished 'build-js' after 3.16 s
[17:15:06] Starting 'tests'...


  ․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․

  74 passing (1s)
  2 pending

[17:15:09] Finished 'tests' after 2.49 s
[17:15:09] Finished 'test' after 5.65 s
```

Thanks!
